support/agent/deal_client_agent.go

```diff
@@ -0,0 +1,206 @@
+}
+
+type DealClientAgent struct {
+	DealCount int
+
+	account    address.Address
+	config     DealClientConfig
+	dealEvents *RateIterator
+	rnd        *rand.Rand
+
+	// tracks funds expected to be locked for client deal payment
+	expectedMarketBalance abi.TokenAmount
+}
+
+func AddDealClientsForAccounts(s SimState, accounts []address.Address, seed int64, config DealClientConfig) []*DealClientAgent {
+	rnd := rand.New(rand.NewSource(seed))
+	var agents []*DealClientAgent
+	for _, account := range accounts {
+		agent := NewDealClientAgent(account, rnd.Int63(), config)
+		agents = append(agents, agent)
+		s.AddAgent(agent)
+	}
+	return agents
+}
+
+func NewDealClientAgent(account address.Address, seed int64, config DealClientConfig) *DealClientAgent {
+	rnd := rand.New(rand.NewSource(seed))
+	return &DealClientAgent{
+		account:               account,
+		config:                config,
+		rnd:                   rnd,
+		expectedMarketBalance: big.Zero(),
+		dealEvents:            NewRateIterator(config.DealRate, rnd.Int63()),
+	}
+}
+
+func (dca *DealClientAgent) Tick(s SimState) ([]message, error) {
+	var providers []DealProvider
+
+	// aggregate all deals into one message
+	if err := dca.dealEvents.Tick(func() error {
+		provider := s.ChooseDealProvider()
+		providers = append(providers, provider)
+		if err := dca.createDeal(s, provider); err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	// add message to update market balance if necessary
+	messages := dca.updateMarketBalance()
+
+	return messages, nil
+}
+
+func (dca *DealClientAgent) updateMarketBalance() []message {
+	if dca.expectedMarketBalance.GreaterThanEqual(dca.config.MinMarketBalance) {
+		return []message{}
+	}
+
+	balanceToAdd := big.Sub(dca.config.MaxMarketBalance, dca.expectedMarketBalance)
+
+	return []message{{
+		From:   dca.account,
+		To:     builtin.StorageMarketActorAddr,
+		Value:  balanceToAdd,
+		Method: builtin.MethodsMarket.AddBalance,
+		Params: &dca.account,
+		ReturnHandler: func(v SimState, msg message, ret cbor.Marshaler) error {
+			dca.expectedMarketBalance = dca.config.MaxMarketBalance
+			return nil
+		},
+	}}
+}
+
+// Create a proposal
+// Return false if provider if deal can't be performed because client or provider lacks funds
+func (dca *DealClientAgent) createDeal(s SimState, provider DealProvider) error {
+	pieceCid, err := dca.generatePieceCID()
+	if err != nil {
+		return err
+	}
+
+	pieceSize := dca.config.MinPieceSize +
+		uint64(dca.rnd.Int63n(int64(dca.config.MaxPieceSize-dca.config.MinPieceSize)))
+
+	// round to next power of two
+	if bits.OnesCount64(pieceSize) > 1 {
+		pieceSize = 1 << bits.Len64(pieceSize)
+	}
+
+	providerCollateral, err := calculateProviderCollateral(s, abi.PaddedPieceSize(pieceSize))
+	if err != nil {
+		return err
+	}
+
+	// if provider does not have enough collateral, just skip this deal
+	if provider.AvailableCollateral().LessThan(providerCollateral) {
```

Comment:
Simulating failures is interesting so removing this check is definitely something we should configure later.  I'm mulling over what is simpler, removing this check and letting failures happen (less code) or keeping this check and having things succeed by default.  Probably the current approach is simpler since failure modes are somewhat simulation implementation dependent and correct protocol execution is not.  Just bringing this up as I'm curious to hear the thinking that got you here.

Reply:
I agree that at some point we should cover failures. We already support failures in ConfirmSectorProofsValid (which appear naturally in the upcoming CC upgrade PR when both faults and upgrades are allowed at a sufficient rate). This currently works because ConfirmSectorProofsValid occurs in cron, and doesn't cause the cron message to fail.

Some design will be required to support messages that fail. Specifically do we need a way to expect failure, or do we simply log failures and continue? Adding an expected exit code to messages wouldn't be hard, and might be a reasonable thing to do in this case. I think I'd prefer to save it for another PR though.